src/osw/express.py

```diff
@@ -0,0 +1,801 @@
+    using the credentials_fp."""
+
+    class Config:
+        arbitrary_types_allowed = True
+
+    def open(self, mode: str = "r", **kwargs):
+        kwargs["mode"] = mode
+        return open(self.path, **kwargs)
+
+    def close(self):
+        self.file.close()
+
+    def read(self, *args, **kwargs):
+        return self.file.read(*args, **kwargs)
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.close()
+        if self.delete_after_use and self.path.exists():
+            self.path.unlink()
+
+    def process_init_data(self, data: Dict[str, Any]) -> Dict[str, Any]:
+        # Get the field default values (attribute defaults)
+        attr_def_vals = {
+            field_name: field.default
+            for field_name, field in self.__class__.__fields__.items()
+        }
+        # Set the default values for the attributes, if no value was passed
+        for key, value in attr_def_vals.items():
+            if data.get(key) is None:
+                data[key] = value
+        # Do replacements
+        if data.get("credentials_fp") is None:
+            data["credentials_fp"] = credentials_fp_default.get_default()
+        if not data.get("credentials_fp").parent.exists():
+            data["credentials_fp"].parent.mkdir(parents=True)
+        if data.get("credential_manager") is None:
+            data["credential_manager"] = CredentialManager()
+            if data.get("credentials_fp").exists():
+                data["credential_manager"] = CredentialManager(
+                    cred_filepath=data.get("credentials_fp")
+                )
+        return data
+
+
+class DownloadFileResult(FileResult, LocalFileController):
+    """A specific result object to describe the result of downloading a file from an
+    OSL instance."""
+
+    url_or_title: str
+    """The URL or full page title of the WikiFile page to download the file from."""
+    target_dir: Optional[Union[str, Path]] = None
+    """The target directory to download the file to. If None, the current working
+    will be used."""
+    target_fn: Optional[str] = None
+    """The target filename to save the file as. If None, the filename will be taken
+    from the URL or title."""
+    target_fp: Optional[Union[str, Path]] = None
+    """The target filepath to save the file to. If None, the file will be saved to the
+    target directory with the target filename."""
+    overwrite: bool = False
+    """If True, the file will be overwritten if it already exists. If False, the file
+    will not be downloaded if it already exists."""
+    use_cached: bool = False
+    """If True, the file will be reloaded from the cache. If False, the file will be
+    reloaded from the server. This option is useful if you are debugging code and
+    don't want to reload the file from the server every time."""
+
+    class Config:
+        arbitrary_types_allowed = True
+
+    def __init__(self, url_or_title, **data):
```

Comment:
Note: in pydantic v2 we can use https://docs.pydantic.dev/latest/api/base_model/#pydantic.BaseModel.model_post_init

Reply:
Should this be replaced?

Reply:
Not yet, since currently all is based on v1